plugins/net_plugin/include/eosio/net_plugin/auto_bp_peering.hpp

```diff
@@ -130,6 +232,152 @@ class bp_connection_manager {
+            const auto& [host, port, type] = net_utils::split_host_port_type(addr);
+            return !host.empty() && !port.empty();
+         };
+         const gossip_bp_peers_message::bp_peer* prev = nullptr;
+         size_t num_per_producer = 0;
+         for (const auto& peer : msg.peers) {
+            if (peer.producer_name.empty())
+               return false; // invalid bp_peer data
+            if (!valid_address(peer.server_address))
+               return false; // invalid address
+            if (prev != nullptr) {
+               if (prev->producer_name == peer.producer_name) {
+                  ++num_per_producer;
+                  if (num_per_producer > max_bp_peers_per_producer)
+                     return false; // more than allowed per producer
+                  if (prev->server_address == peer.server_address)
+                     return false; // duplicate entries not allowed
+               } else if (prev->producer_name > peer.producer_name) {
+                  return false; // required to be sorted
+               } else {
+                  num_per_producer = 0;
+               }
+            }
+            prev = &peer;
+         }
+      }
+
+      const controller& cc = self()->chain_plug->chain();
+      bool fatal_error = false;
+      auto is_peer_key_valid = [&](const gossip_bp_peers_message::bp_peer& peer) -> bool {
+         try {
+            if (peer.sig.is_webauthn()) {
+               fc_dlog(self()->get_logger(), "Peer ${p} signature is webauthn, not allowed.", ("p", peer.producer_name));
+               fatal_error = true;
+               return false;
+            }
+            std::optional<peer_info_t> peer_info = cc.get_peer_info(peer.producer_name);
+            if (peer_info && peer_info->key) {
+               constexpr bool check_canonical = false;
+               public_key_type pk(peer.sig, peer.digest(), check_canonical);
+               if (pk != *peer_info->key) {
+                  fc_dlog(self()->get_logger(), "Recovered peer key did not match on-chain ${p}, recovered: ${pk} != expected: ${k}",
+                          ("p", peer.producer_name)("pk", pk)("k", *peer_info->key));
+                  return false;
+               }
+            } else { // unknown key
+               // ok, might have just been deleted or dropped out of top ranking
+               fc_dlog(self()->get_logger(), "Failed to find peer key ${p}", ("p", peer.producer_name));
+               return false;
+            }
+         } catch (fc::exception& e) {
+            fc_dlog(self()->get_logger(), "Exception recovering peer key ${p}, error: ${e}", ("p", peer.producer_name)("e", e.to_detail_string()));
+            fatal_error = true;
+            return false; // invalid key
+         }
+         return true;
+      };
+
+      fc::lock_guard g(gossip_bps.mtx);
+      auto& sig_idx = gossip_bps.index.get<by_sig>();
+      for (auto i = msg.peers.begin(); i != msg.peers.end() && !fatal_error;) {
```

Comment:
I don't think we need to exit the loop early if `fatal_error` is true (it is after all an optimization for a very unlikely case).

And if we don't, we can use `std::erase` instead of that for loop which makes the code clearer I feel.

Reply:
nit: Maybe `invalid_message` would be more descriptive than `fatal_error`.

Reply:
I think it is important to exit early on a bad `public_key`. Seems like those could be crafted to be expensive to evaluate. As soon as one is determined to be invalid, don't bother trying to verify signature on any others.

Reply:
https://github.com/AntelopeIO/spring/pull/1265/commits/75c63e39f526c5ffe5d2de44e95fe98f3c521905

Reply:
> Seems like those could be crafted to be expensive to evaluate.
I don't think it is a very likely attack vector from one of the top-50 producers, but OK.

Reply:
It could be anyone that is connected to the node.

Reply:
That aspect is what I've been thinking on some and why suggested disallowing SIG_WA -- those have the most opportunity to create pain on the very first signature in the message.

I have been trying to think of what else could be problematic here. I don't see a limit for the size of `gossip_bp_peers_message` which means you could have as many peers as you can stuff in 8MB (or until hitting max array elements). So someone can flood `gossip_bp_peers_message`s with ~110k peers each taking ~40µs to recover key leading to over 4 seconds of work here?

Reply:
Only `max_bp_peers_per_producer` (4) per producer is allowed. The number allowed here will be limited to 4x(max-allowed-by-contract) which is currently 200 (4x50).

Reply:
That limit is enforced prior to checking the signatures though. So as an attacker why can't I just submit a message with names `a`, `b`, `c`, etc all with valid signatures.

Reply:
`a`, `b`, `c`, have to be in the top 50 to eval the signature.
https://github.com/AntelopeIO/spring/blob/75c63e39f526c5ffe5d2de44e95fe98f3c521905/plugins/net_plugin/include/eosio/net_plugin/auto_bp_peering.hpp#L281-L282

Reply:
> It could be anyone that is connected to the node.

I thought the peer gossip was restricted to the top-50 producers. Maybe we should ignore `peer_gossip` messages if they don't come from one of the top-50 producers.

Reply:
I take that back. We do send a full bp peer message to the peers sent to us. We only validate the initial one, after that we trust the info that gossip peer sends up.

Reply:
If the message is not from one of the top-50 we know about then we will not consider that a valid message and will disconnect.

Reply:
So this means that the attack with invalid signatures is not very likely, right. Not advocating for a change, but just that these messages are among a trusted group (all trusted to see the IP of all block producers, so if there was a rogue element it could probably do more damage by ddos'ing producers as they are scheduled to produce than by forging public keys that take time to validate).

Reply:
Well anyone could send a crafted `gossip_bp_peers_message` we have to validate to know it is from a valid producer. 

This conversation did make me realize that it was a bit too strict. Just because a peer is no longer in the top producers does not mean we should `fatal` disconnect from them.

Reply:
> Well anyone could send a crafted `gossip_bp_peers_message` we have to validate to know it is from a valid producer.

Hum, I wonder if the whole message (the `std::vector<bp_peer>` + sender's name) should not be signed by the peer sending it. Then the first thing we could do is validate that the signature matches for one  connection IP, so we authenticate that it comes from one of the current `top-50` producers.

Reply:
We could certainly do that. Not sure it is worth the extra processing. Open to what others think.